src/nvapi/nvapi_d3d12_device.cpp

```diff
@@ -27,180 +91,86 @@ namespace dxvk {
-
-        return CreateCubinComputeShaderEx(device, cubinData, cubinSize, blockX, blockY, blockZ, 0 /* smemSize */, shaderName, pShader);
+    bool NvapiD3d12Device::CreateCubinComputeShaderWithName(const void* cubinData, NvU32 cubinSize, NvU32 blockX, NvU32 blockY, NvU32 blockZ, const char* shaderName, NVDX_ObjectHandle* pShader) {
+        return CreateCubinComputeShaderEx(cubinData, cubinSize, blockX, blockY, blockZ, 0 /* smemSize */, shaderName, pShader);
     }
 
-    bool NvapiD3d12Device::CreateCubinComputeShaderEx(ID3D12Device* device, const void* cubinData, NvU32 cubinSize, NvU32 blockX, NvU32 blockY, NvU32 blockZ, NvU32 smemSize, const char* shaderName, NVDX_ObjectHandle* pShader) {
-        auto cubinDevice = GetCubinDevice(device);
-        if (cubinDevice == nullptr)
-            return false;
-
-        Com<ID3D12DeviceExt> deviceExt;
-        if (FAILED(cubinDevice->QueryInterface(IID_PPV_ARGS(&deviceExt))))
+    bool NvapiD3d12Device::CreateCubinComputeShaderEx(const void* cubinData, NvU32 cubinSize, NvU32 blockX, NvU32 blockY, NvU32 blockZ, NvU32 smemSize, const char* shaderName, NVDX_ObjectHandle* pShader) {
+        if (!m_vkd3dDevice || !m_supportsNvxBinaryImport)
             return false;
 
-        if (FAILED(deviceExt->CreateCubinComputeShaderWithName(cubinData, cubinSize, blockX, blockY, blockZ, shaderName, reinterpret_cast<D3D12_CUBIN_DATA_HANDLE**>(pShader))))
+        if (FAILED(m_vkd3dDevice->CreateCubinComputeShaderWithName(cubinData, cubinSize, blockX, blockY, blockZ, shaderName, reinterpret_cast<D3D12_CUBIN_DATA_HANDLE**>(pShader))))
             return false;
 
         std::scoped_lock lock(m_cubinSmemMutex);
         m_cubinSmemMap.emplace(*pShader, smemSize);
         return true;
     }
 
-    bool NvapiD3d12Device::DestroyCubinComputeShader(ID3D12Device* device, NVDX_ObjectHandle shader) {
-        auto cubinDevice = GetCubinDevice(device);
-        if (cubinDevice == nullptr)
+    bool NvapiD3d12Device::DestroyCubinComputeShader(NVDX_ObjectHandle shader) {
+        if (!m_vkd3dDevice || !m_supportsNvxBinaryImport)
             return false;
 
-        if (FAILED(cubinDevice->DestroyCubinComputeShader(reinterpret_cast<D3D12_CUBIN_DATA_HANDLE*>(shader))))
+        if (FAILED(m_vkd3dDevice->DestroyCubinComputeShader(reinterpret_cast<D3D12_CUBIN_DATA_HANDLE*>(shader))))
             return false;
 
         std::scoped_lock lock(m_cubinSmemMutex);
         m_cubinSmemMap.erase(shader);
         return true;
     }
 
-    bool NvapiD3d12Device::GetCudaTextureObject(ID3D12Device* device, D3D12_CPU_DESCRIPTOR_HANDLE srvHandle, D3D12_CPU_DESCRIPTOR_HANDLE samplerHandle, NvU32* cudaTextureHandle) {
-        auto cubinDevice = GetCubinDevice(device);
-        if (cubinDevice == nullptr)
+    bool NvapiD3d12Device::GetCudaTextureObject(D3D12_CPU_DESCRIPTOR_HANDLE srvHandle, D3D12_CPU_DESCRIPTOR_HANDLE samplerHandle, NvU32* cudaTextureHandle) const {
+        if (!m_vkd3dDevice || !m_supportsNvxBinaryImport)
```

Comment:
As with D3D11, technically it's `VK_NVX_image_view_handle` that we care about here. But feel free to keep this, in practice it probably doesn't matter as long as NVK doesn't decide to implement one extension without implementing the other.

Reply:
I’ve put this on my todo list for a follow up pr, also for d3d11.